test/unittests/libp2p/net.cpp

```diff
@@ -430,6 +450,144 @@ BOOST_AUTO_TEST_CASE(noteActiveNodeReplacesNodeInFullBucketWhenEndpointChanged)
+    TestUDPSocketHost nodeSocketHost{30500};
+    nodeSocketHost.start();
+    uint16_t nodePort = nodeSocketHost.port;
+
+    // add a node to node table, initiating PING
+    auto nodeEndpoint = NodeIPEndpoint{bi::address::from_string("127.0.0.1"), nodePort, nodePort};
+    auto nodeKeyPair = KeyPair::create();
+    auto nodePubKey = nodeKeyPair.pub();
+    nodeTable->addNode(Node{nodePubKey, nodeEndpoint});
+
+    // send PONG
+    Pong pong(nodeTable->m_hostNode.endpoint);
+    pong.sign(nodeKeyPair.secret());
+
+    nodeSocketHost.socket->send(pong);
+
+    // wait for PONG to be received and handled
+    nodeTable->packetReceived.get_future().wait();
+
+    auto addedNode = nodeTable->m_allNodes.find(nodePubKey);
+    BOOST_REQUIRE(addedNode != nodeTable->m_allNodes.end());
+    BOOST_CHECK_EQUAL(addedNode->second->lastPongReceivedTime, 0);
+}
+
+BOOST_AUTO_TEST_CASE(validPong)
+{
+    // NodeTable sending PING
+    TestNodeTableHost nodeTableHost(0);
+    nodeTableHost.start();
+    auto& nodeTable = nodeTableHost.nodeTable;
+
+    // socket receiving PING
+    TestUDPSocketHost nodeSocketHost{30500};
+    nodeSocketHost.start();
+    uint16_t nodePort = nodeSocketHost.port;
+
+    // add a node to node table, initiating PING
+    auto nodeEndpoint = NodeIPEndpoint{bi::address::from_string("127.0.0.1"), nodePort, nodePort};
+    auto nodeKeyPair = KeyPair::create();
+    auto nodePubKey = nodeKeyPair.pub();
+    nodeTable->addNode(Node{nodePubKey, nodeEndpoint});
+
+    // handle received PING
+    auto pingDataReceived = nodeSocketHost.packetReceived.get_future().get();
+    auto pingDatagram =
+        DiscoveryDatagram::interpretUDP(bi::udp::endpoint{}, dev::ref(pingDataReceived));
+    auto ping = dynamic_cast<PingNode const&>(*pingDatagram);
+
+    // send PONG
+    Pong pong(nodeTable->m_hostNode.endpoint);
+    pong.echo = ping.echo;
+    pong.sign(nodeKeyPair.secret());
+    nodeSocketHost.socket->send(pong);
+
+    // wait for PONG to be received and handled
+    nodeTable->packetReceived.get_future().wait();
+
+    auto addedNode = nodeTable->m_allNodes.find(nodePubKey);
+    BOOST_REQUIRE(addedNode != nodeTable->m_allNodes.end());
+    BOOST_CHECK(addedNode->second->lastPongReceivedTime > 0);
+}
+
+BOOST_AUTO_TEST_CASE(pingTimeout)
+{
+    // NodeTable sending PING
+    TestNodeTableHost nodeTableHost(0);
+    nodeTableHost.start();
+    auto& nodeTable = nodeTableHost.nodeTable;
+
+    // socket receiving PING
+    TestUDPSocketHost nodeSocketHost{30500};
+    nodeSocketHost.start();
+    uint16_t nodePort = nodeSocketHost.port;
+
+    // add a node to node table, initiating PING
+    auto nodeEndpoint = NodeIPEndpoint{bi::address::from_string("127.0.0.1"), nodePort, nodePort};
+    auto nodeKeyPair = KeyPair::create();
+    auto nodePubKey = nodeKeyPair.pub();
+    nodeTable->addNode(Node{nodePubKey, nodeEndpoint});
+
+    this_thread::sleep_for(DiscoveryDatagram::c_timeToLive + std::chrono::seconds(5));
```

Comment:
Not sure if it's acceptable for a test to sleep for 25 seconds... I'll probably try to make it configurable